src/hardware/driver/output.rs

```diff
@@ -0,0 +1,176 @@
+    iir: iir::IIR<f32>,
+    relay: Relay<I2C>,
+}
+
+impl<I2C, E> Output<I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    // 1 A/s current ramp-up
+    const RAMP_STEP: f32 = 1.0; // will be 1 mA current steps
+    const RAMP_DELAY: fugit::MillisDuration<u64> =
+        fugit::MillisDurationU64::millis(10); // will be current steps every 1 ms
+
+    pub fn new(
+        gpio: &'static spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
+        channel: Channel,
+    ) -> Self {
+        Output {
+            iir: iir::IIR::new(0., -i16::MAX as _, i16::MAX as _), // Todo: sensible defaults for driver (this is for stabilizer now)
+            relay: Relay::new(gpio, channel),
+        }
+    }
+}
+
+pub mod sm {
+    use super::*;
+    statemachine! {
+        transitions: {
+            // Enable sequence
+            *Disabled + Enable / engage_k0 = EnableWaitK0,
+            EnableWaitK0 + RelayDone / disengage_k1 = EnableWaitK1,
+            EnableWaitK1 + RelayDone = RampCurrent,
+            RampCurrent + CurrentStep / increment_current = RampCurrent,
+            RampCurrent + CurrentFinal = Enabled,
+
+            // Abort transitions
+            EnableWaitK0 | EnableWaitK1 | RampCurrent + Disable = Abort,
+            Abort + RelayDone / engage_k1_and_hold_iir = DisableWaitK1,
+
+            // Disable sequence
+            Enabled + Disable / engage_k1_and_hold_iir = DisableWaitK1,
+            DisableWaitK1 + RelayDone / disengage_k0_and_reset_iir = DisableWaitK0,
+            DisableWaitK0 + RelayDone = Disabled,
+
+            Disabled + Disable = Disabled,
+            Enabled + Enable = Enabled
+        }
+    }
+}
+
+impl<I2C, E> sm::StateMachineContext for Output<I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    fn increment_current(&mut self) {
+        self.iir.y_offset += Output::<I2C>::RAMP_STEP;
+    }
+
+    fn engage_k0(&mut self) {
+        self.relay.engage_k0()
+    }
+
+    fn disengage_k0_and_reset_iir(&mut self) {
+        self.iir.y_offset = 0.;
+        self.iir.ba = [0., 0., 0., 0., 0.];
+        self.relay.disengage_k0()
+    }
+
+    fn disengage_k1(&mut self) {
+        self.relay.disengage_k1()
+    }
+
+    // set K1 to lower position and output current to zero
+    fn engage_k1_and_hold_iir(&mut self) {
+        self.iir.y_offset = 0.;
+        self.iir.ba = [0., 0., 0., 1., 0.];
+        self.relay.engage_k1()
+    }
+}
+
+impl<I2C, E> sm::StateMachine<Output<I2C>>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    /// Start enabling sequence. Returns `Some(relay delay)` or None if we started an Abort
+    /// or an error if SM is in a state where we can't enable.
+    pub fn enable(
+        &mut self,
+    ) -> Result<Option<fugit::MillisDuration<u64>>, sm::Error> {
+        self.process_event(sm::Events::Enable)?;
+        if *self.state() != sm::States::Abort {
```

Comment:
Look at the `process_event()` signature!

```suggestion
        if self.process_event(sm::Events::Enable)? != sm::States::Abort {
```